### PR TITLE
[FEATURE] Add a subset string editor provider concept and registry 

### PR DESCRIPTION
**Pull request by rouault**
The aim of this work is to be able to provide custom subset string editor
GUI according to the layer. Typically, so that a WFS layer uses the same
editor than in its select source, or that a plugin can provide a custom
editor.

* Add QgsSubsetStringEditorInterface: abstract interface to define a dialog
  that can edit a subset string
* Make QgsQueryBuilder implement QgsSubsetStringEditorInterface
* Add QgsSubsetStringEditorProvider: interface for thos who want to provide
  a dialog to edit a subset string.
* Add QgsSubsetStringEditorProviderRegistry: keeps a list of subset string
  editor providers. Transposed from QgsDataItemGuiProviderRegistry
* Add QgsGui::subsetStringEditorProviderRegistry()

Make 'Add Filter' context menu entry and 'Edit query' in vector layer
properties dialog use the subset string provider registry

WFS provider: declare its QgsSQLComposerDialog implementation as a
subset string editor provider

### CODE REVIEW

File: src/gui/qgssqlcomposerdialog.cpp [hunk 1]
```diff
@@ -21,14 +21,20 @@ email                : even.rouault at spatialys.com
...
 
 #include <QMessageBox>
 #include <QKeyEvent>
 
 #include <Qsci/qscilexer.h>
 
-QgsSQLComposerDialog::QgsSQLComposerDialog( QWidget *parent, Qt::WindowFlags fl )
-  : QDialog( parent, fl )
+QgsSQLComposerDialog::QgsSQLComposerDialog( QWidget *parent, Qt::WindowFlags fl ) :
+  QgsSQLComposerDialog( nullptr, parent, fl )
```
Comment (nyalldawson) [inline src/gui/qgssqlcomposerdialog.cpp:32]:
```suggestion
QgsSQLComposerDialog::QgsSQLComposerDialog( QWidget *parent, Qt::WindowFlags fl )
  : QgsSQLComposerDialog( nullptr, parent, fl )
```

File: src/gui/vector/qgsvectorlayerproperties.cpp [hunk 2]
```diff
@@ -844,24 +846,24 @@ void QgsVectorLayerProperties::urlClicked( const QUrl &url )
 void QgsVectorLayerProperties::pbnQueryBuilder_clicked()
 {
   // launch the query builder
-  QgsQueryBuilder *qb = new QgsQueryBuilder( mLayer, this );
+  auto *dialog = QgsGui::subsetStringEditorProviderRegistry()->createDialog( mLayer, this );
```
Comment (nyalldawson) [inline src/gui/vector/qgsvectorlayerproperties.cpp:849]: Can you skip the auto here please? It's not clear from the surrounds what the return type is

File: src/providers/wfs/qgswfsprovidergui.cpp [hunk 3]
```diff
@@ -35,6 +36,29 @@ class QgsWfsSourceSelectProvider : public QgsSourceSelectProvider
...
+class QgsWfsSubsetStringEditorProvider: public QgsSubsetStringEditorProvider
+{
+  public:
+
+    QString providerKey() const override { return QgsWFSProvider::WFS_PROVIDER_KEY; }
+
+    bool canHandleLayer( QgsVectorLayer *layer ) const override
+    {
+      QgsDataProvider *provider = layer->dataProvider();
+      return dynamic_cast<QgsWFSProvider *>( provider ) != nullptr;
```
Comment (nyalldawson) [inline src/providers/wfs/qgswfsprovidergui.cpp:49]:
```suggestion
      return static_cast< bool >( dynamic_cast<QgsWFSProvider *>( provider ) );
```

File: src/providers/wfs/qgswfsprovidergui.cpp [hunk 4]
```diff
@@ -35,6 +36,29 @@ class QgsWfsSourceSelectProvider : public QgsSourceSelectProvider
...
+    }
+
+    QgsSubsetStringEditorInterface *createDialog( QgsVectorLayer *layer, QWidget *parent, Qt::WindowFlags fl ) override
+    {
+      QgsDataProvider *provider = layer->dataProvider();
+      QgsWFSProvider *wfsProvider = dynamic_cast<QgsWFSProvider *>( provider );
+      if ( wfsProvider == nullptr )
```
Comment (nyalldawson) [inline src/providers/wfs/qgswfsprovidergui.cpp:56]:
```suggestion
      if ( !wfsProvider )
```

File: src/gui/qgssqlcomposerdialog.cpp (hunk 1 above)
Comment (rouault) [reply src/gui/qgssqlcomposerdialog.cpp]: done

File: src/providers/wfs/qgswfsprovidergui.cpp (hunk 3 above)
Comment (rouault) [reply src/providers/wfs/qgswfsprovidergui.cpp]: done

File: src/providers/wfs/qgswfsprovidergui.cpp (hunk 4 above)
Comment (rouault) [reply src/providers/wfs/qgswfsprovidergui.cpp]: done

File: src/gui/vector/qgsvectorlayerproperties.cpp (hunk 2 above)
Comment (rouault) [reply src/gui/vector/qgsvectorlayerproperties.cpp]: done

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@rouault I'm seeing the new test crash randomly on some PRs - e.g. https://travis-ci.org/github/qgis/QGIS/builds/744328582

**Comment by rouault** [comment]
> @rouault I'm seeing the new test crash randomly on some PRs - e.g. https://travis-ci.org/github/qgis/QGIS/builds/744328582

should be addressed by https://github.com/qgis/QGIS/pull/40152

**Pull request merged**